File: backend/app/services/auth_service.py
from datetime import datetime, timedelta, timezone
from typing import Tuple, Dict, Any, Optional
from bson import ObjectId
from pymongo.database import Database
from pymongo.errors import DuplicateKeyError
from fastapi import HTTPException, status

from app.config import settings
from app.schemas.auth import (
    UserRegisterRequest,
    UserLoginRequest,
    UserResponse,
)
from app.utils.security import hash_password, verify_password, create_access_token

_indexes_initialized = False


def ensure_user_indexes(db: Database):
    """Ensure unique indexes for mobile and optional email are created on users collection."""
    global _indexes_initialized
    if _indexes_initialized or db is None:
        return
    try:
        # Unique index on mobile
        db.users.create_index("mobile", unique=True, name="uniq_mobile")
        
        # Unique index on email with partial filter for non-null/non-empty strings
        db.users.create_index(
            "email",
            unique=True,
            partialFilterExpression={"email": {"$type": "string", "$gt": ""}},
            name="uniq_email_partial"
        )

        # Ensure default system admin user exists
        admin_doc = db.users.find_one({"role": "admin"})
        if not admin_doc:
            now_utc = datetime.now(timezone.utc)
            db.users.insert_one({
                "name": "SJ Jewellers Admin",
                "mobile": "9999999999",
                "email": "admin@sjjewelers.com",
                "password_hash": hash_password("admin123"),
                "role": "admin",
                "account_status": "active",
                "kyc_status": "verified",
                "profile_completed": True,
                "created_at": now_utc,
                "updated_at": now_utc,
            })

        _indexes_initialized = True
    except Exception:
        # Non-fatal during startup; will retry on next operation
        pass


def _normalize_mobile(raw_mobile: str) -> Tuple[str, list]:
    """Helper to clean and get mobile search variants."""
    clean = (raw_mobile or "").strip()
    digits = "".join(filter(str.isdigit, clean))
    
    # Strip any redundant leading 91 or 0 prefixes while length > 10
    while len(digits) > 10:
        if digits.startswith("91"):
            digits = digits[2:]
        elif digits.startswith("0"):
            digits = digits[1:]
        else:
            break

    if len(digits) == 10:
        clean = digits

    variants = list(set([clean, f"+91{clean}", f"91{clean}", (raw_mobile or "").strip()]))
    return clean, variants




def register_user(db: Database, data: UserRegisterRequest) -> UserResponse:
    """Register a new customer account securely in MongoDB."""
    if db is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Database service unavailable",
        )
    
    ensure_user_indexes(db)
    clean_mobile, mobile_variants = _normalize_mobile(data.mobile)

    # Check for existing mobile
    if db.users.find_one({"mobile": {"$in": mobile_variants}}):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Mobile number already registered",
        )

    # Check for existing email if provided
    if data.email:
        if db.users.find_one({"email": data.email}):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Email already registered",
            )

    now = datetime.now(timezone.utc)
    hashed_pwd = hash_password(data.password)

    user_doc = {
        "name": data.name,
        "mobile": clean_mobile,
        "email": data.email,
        "password_hash": hashed_pwd,
        "role": "customer",
        "account_status": "active",
        "kyc_status": "pending",
        "profile_completed": False,
        "created_at": now,
        "updated_at": now,
    }

    try:
        result = db.users.insert_one(user_doc)
    except DuplicateKeyError as e:
        err_msg = str(e)
        if "email" in err_msg:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Email already registered")
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Mobile number already registered")

    user_id = str(result.inserted_id)
    return UserResponse(
        id=user_id,
        name=user_doc["name"],
        mobile=user_doc["mobile"],
        email=user_doc["email"],
        role=user_doc["role"],
        account_status=user_doc["account_status"],
        kyc_status=user_doc["kyc_status"],
        profile_completed=False,
        created_at=user_doc["created_at"],
    )


def login_user(db: Database, data: UserLoginRequest) -> Tuple[UserResponse, str]:
    """Authenticate customer credentials and return user profile + JWT token."""
    if db is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Database service unavailable",
        )

    ensure_user_indexes(db)

    # Lookup user by mobile, mobile variants, email, or name
    raw_ident = (data.mobile or "").strip()
    clean_mobile, mobile_variants = _normalize_mobile(raw_ident)
    ident_variants = list(set([raw_ident, clean_mobile] + mobile_variants))

    query_conditions = [
        {"mobile": {"$in": ident_variants}},
        {"email": raw_ident.lower()},
        {"name": raw_ident},
    ]
    if raw_ident.lower() in ["admin", "administrator"]:
        query_conditions.append({"role": "admin"})

    query = {"$or": query_conditions}
    user_doc = db.users.find_one(query)
    
    # Generic error to prevent user enumeration
    if not user_doc or not verify_password(data.password, user_doc.get("password_hash", "")):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid mobile number or password",
            headers={"WWW-Authenticate": "Bearer"},
        )

    # Check account status
    account_status = user_doc.get("account_status", "active")
    if account_status == "banned":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Your account has been banned",
        )
    elif account_status == "suspended":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Your account is currently suspended",
        )
    elif account_status != "active":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=f"Account is {account_status}. Please contact support.",
        )

    user_id = str(user_doc["_id"])
    token_payload = {
        "sub": user_id,
        "role": user_doc.get("role", "customer"),
        "mobile": user_doc.get("mobile"),
    }
    
    access_token = create_access_token(token_payload)

    # Check profile completion state
    profile_completed = user_doc.get("profile_completed", False)
    if not profile_completed:
        # Check if user profile is already populated with address and details
        prof = db.profiles.find_one({"user_id": user_id})
        if prof and prof.get("full_name") and prof.get("address"):
            profile_completed = True

    user_response = UserResponse(
        id=user_id,
        name=user_doc.get("name", ""),
        mobile=user_doc.get("mobile", ""),
        email=user_doc.get("email"),
        role=user_doc.get("role", "customer"),
        account_status=account_status,
        kyc_status=user_doc.get("kyc_status", "pending"),
        profile_completed=profile_completed,
        created_at=user_doc.get("created_at"),
    )

    return user_response, access_token


def format_user_response(user_doc: Dict[str, Any]) -> UserResponse:
    """Format raw MongoDB user document into a safe UserResponse schema."""
    return UserResponse(
        id=str(user_doc.get("_id", user_doc.get("id"))),
        name=user_doc.get("name", ""),
        mobile=user_doc.get("mobile", ""),
        email=user_doc.get("email"),
        role=user_doc.get("role", "customer"),
        account_status=user_doc.get("account_status", "active"),
        kyc_status=user_doc.get("kyc_status", "pending"),
        profile_completed=user_doc.get("profile_completed", False),
        created_at=user_doc.get("created_at"),
    )

File: backend/src/services/authService.js
import { v4 as uuidv4 } from 'uuid';
import { query } from '../config/db.js';
import config from '../config/env.js';
import { hashPassword, verifyPassword, createAccessToken } from '../utils/security.js';
import { normalizeMobile } from '../utils/formatters.js';

export async function sendOtp(rawMobile) {
  const { mobile } = normalizeMobile(rawMobile);
  if (!mobile || mobile.length !== 10) {
    const error = new Error('Please provide a valid 10-digit mobile number');
    error.status = 400;
    throw error;
  }

  const otpCode = config.devOtp;
  const now = new Date();
  const expiresAt = new Date(now.getTime() + 10 * 60 * 1000); // 10 minutes

  const id = uuidv4();
  await query(
    `INSERT INTO otps (id, mobile, otp, created_at, expires_at)
     VALUES (?, ?, ?, NOW(), ?)
     ON DUPLICATE KEY UPDATE otp = VALUES(otp), expires_at = VALUES(expires_at)`,
    [id, mobile, otpCode, expiresAt]
  );

  return { mobile, otpCode };
}

export async function verifyOtp(rawMobile, enteredOtp) {
  const { mobile } = normalizeMobile(rawMobile);
  const cleanOtp = (enteredOtp || '').toString().trim();

  if (!cleanOtp) {
    const error = new Error('OTP is required');
    error.status = 400;
    throw error;
  }

  if (cleanOtp === config.devOtp) {
    return true;
  }

  const rows = await query('SELECT * FROM otps WHERE mobile = ? LIMIT 1', [mobile]);
  if (rows.length > 0) {
    const record = rows[0];
    if (record.otp === cleanOtp) {
      const expiresAt = new Date(record.expires_at);
      if (expiresAt > new Date()) {
        return true;
      }
    }
  }

  const error = new Error('Invalid or expired OTP');
  error.status = 400;
  throw error;
}

export async function registerUser({ name, mobile: rawMobile, email, password }) {
  const { mobile, variants } = normalizeMobile(rawMobile);

  if (!mobile || mobile.length !== 10) {
    const error = new Error('Please provide a valid 10-digit mobile number');
    error.status = 400;
    throw error;
  }

  if (!password || password.length < 6) {
    const error = new Error('Password must be at least 6 characters');
    error.status = 400;
    throw error;
  }

  // Check mobile existence
  const placeholders = variants.map(() => '?').join(',');
  const existingMobile = await query(`SELECT id FROM users WHERE mobile IN (${placeholders}) LIMIT 1`, variants);
  if (existingMobile.length > 0) {
    const error = new Error('Mobile number already registered');
    error.status = 400;
    throw error;
  }

  // Check email existence if provided
  const cleanEmail = email ? email.trim().toLowerCase() : null;
  if (cleanEmail) {
    const existingEmail = await query('SELECT id FROM users WHERE email = ? LIMIT 1', [cleanEmail]);
    if (existingEmail.length > 0) {
      const error = new Error('Email already registered');
      error.status = 400;
      throw error;
    }
  }

  const hashedPassword = await hashPassword(password);
  const userId = uuidv4();
  const cleanName = (name || '').trim() || 'Customer';

  await query(
    `INSERT INTO users (id, name, mobile, email, password_hash, role, account_status, kyc_status, profile_completed, created_at, updated_at)
     VALUES (?, ?, ?, ?, ?, 'customer', 'active', 'pending', 0, NOW(), NOW())`,
    [userId, cleanName, mobile, cleanEmail, hashedPassword]
  );

  // Initialize empty profile
  const profileId = uuidv4();
  await query(
    `INSERT INTO profiles (id, user_id, full_name, created_at, updated_at)
     VALUES (?, ?, ?, NOW(), NOW())`,
    [profileId, userId, cleanName]
  );

  const tokenPayload = { sub: userId, role: 'customer', mobile };
  const accessToken = createAccessToken(tokenPayload);

  return {
    user: {
      id: userId,
      name: cleanName,
      mobile,
      email: cleanEmail,
      role: 'customer',
      account_status: 'active',
      kyc_status: 'pending',
      profile_completed: false,
      created_at: new Date().toISOString(),
    },
    accessToken,
  };
}

export async function loginUser({ mobile: rawIdent, password }) {
  const ident = (rawIdent || '').toString().trim();
  const { mobile, variants } = normalizeMobile(ident);
  const allIdents = Array.from(new Set([ident, mobile, ...variants]));

  let user = null;

  // Search by mobile or email or username
  const placeholders = allIdents.map(() => '?').join(',');
  const querySql = `
    SELECT * FROM users 
    WHERE mobile IN (${placeholders}) 
       OR email = ? 
       OR name = ?
       ${ident.toLowerCase() === 'admin' ? "OR role = 'admin'" : ''}
    LIMIT 1
  `;
  const users = await query(querySql, [...allIdents, ident.toLowerCase(), ident]);

  if (users.length > 0) {
    user = users[0];
  }

  if (!user || !(await verifyPassword(password, user.password_hash))) {
    const error = new Error('Invalid mobile number or password');
    error.status = 401;
    throw error;
  }

  if (user.account_status === 'banned') {
    const error = new Error('Your account has been banned');
    error.status = 403;
    throw error;
  }

  if (user.account_status === 'suspended') {
    const error = new Error('Your account is currently suspended');
    error.status = 403;
    throw error;
  }

  if (user.account_status !== 'active') {
    const error = new Error(`Account is ${user.account_status}. Please contact support.`);
    error.status = 403;
    throw error;
  }

  const tokenPayload = {
    sub: user.id,
    role: user.role,
    mobile: user.mobile,
  };
  const accessToken = createAccessToken(tokenPayload);

  return {
    user: {
      id: user.id,
      name: user.name,
      mobile: user.mobile,
      email: user.email,
      role: user.role,
      account_status: user.account_status,
      kyc_status: user.kyc_status,
      profile_completed: Boolean(user.profile_completed),
      created_at: user.created_at,
    },
    accessToken,
  };
}

export function formatUserResponse(user) {
  return {
    id: user.id,
    name: user.name,
    mobile: user.mobile,
    email: user.email,
    role: user.role,
    account_status: user.account_status,
    kyc_status: user.kyc_status,
    profile_completed: Boolean(user.profile_completed),
    created_at: user.created_at,
  };
}

export default {
  sendOtp,
  verifyOtp,
  registerUser,
  loginUser,
  formatUserResponse,
};
